### DISCUSSION TITLE
Prefetch and consumer acknowledgement timeout

### DISCUSSION THREAD

**Discussion by pabermod** [rabbitmq/rabbitmq-server#7250] [Questions] [answered]
Hi, 

I have the following scenario:

Delivery acknowledgement is enabled.

- I send 100 messages
- The consumer has a prefetch of 60 and does the following process for each messages that receives:
  - Process the message (takes about 130 seconds)
  - Confirms the message

This is giving me timeout :

Channel 'ERP(1)' shutdown: 406. PRECONDITION_FAILED - delivery acknowledgement on channel 1 timed out. Timeout value used: 1800000 ms. This timeout value can be configured, see consumers doc guide to learn more

The timeout of the last message is taking into account the time to confirm the previous messages? I don't know if this is a bug or if its the desired behaviour.

Regards

**Comment by mkuratczyk** [comment, marked as answer]
You ask for 60 messages and it takes ~2 minutes to process a single message, so after 30 minutes, you hit the timeout, because you still haven't confirmed ~45 messages. If it takes 2 minutes to process a message, I see no reason for prefetch larger than 1 or 2.

**Reply by pabermod** [reply]
Hi, thanks for your answer. I'm confirming messages one by one. 

I think this behaviour should be documented, because I thought that the timeout was applied for each message and not for all the messages prefetched and is confusing, because in reality each message is confirmed in 2 minutes

**Reply by pabermod** [reply]
I ask for 60 messages but confirm one by one... I think maybe its a bug?

**Reply by mkuratczyk** [reply]
The idea of the timeout it to react to a situation when a consumer receives a message but "never" confirms it (by default "never" means "30 minutes"). This time is measured for all 60 messages that were dispatched to your consumer. It doesn't matter that you acknowledged some of them - you still have many that you haven't acknowledged in the 30 minutes since you received them. Lowering the prefetch is the right solution - you will receive fewer, so all you need to do is to acknowledge any message that you receive, within 30 minutes since you received it (unless you change the timeout but there is no reason to do that, since you expect message processing to take much less than that).

Large prefetch makes sense when the latency to get a message is significant compared to the time it takes to process the message. In such cases, you want to have some additional messages ready to process, while more is being delivered to the consumer. But if it takes 2 minutes to process a message, the time it takes to get the next message is insignificant. So just set the prefetch to 2, so that you can start processing the next message as soon as you finished the last one and that's it.

**Reply by pabermod** [reply]
Thank you Michal, I was not understanding how the prefetch and timeout work as a whole

**Reply by lukebakken** [reply]
Thanks @mkuratczyk this is very informative.
